File: src/main/java/macro303/keschet/core/pieces/Piece.java
package macro303.keschet.core.pieces;

import macro303.board_game.Colour;
import macro303.board_game.IPiece;
import macro303.keschet.Direction;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

/**
 * Created by Macro303 on 2018-02-08.
 */
public abstract class Piece implements IPiece {
	private final int maxDistance;
	@NotNull
	private final String symbol;
	@NotNull
	private final Direction[] validDirections;
	@NotNull
	private Colour colour;

	protected Piece(int maxDistance, @NotNull String symbol, @NotNull Direction[] validDirections) {
		this(Colour.RESET, maxDistance, symbol, validDirections);
	}

	protected Piece(@NotNull Colour colour, int maxDistance, @NotNull String symbol, @NotNull Direction[] validDirections) {
		this.colour = colour;
		this.maxDistance = maxDistance;
		this.symbol = symbol;
		this.validDirections = validDirections;
	}

	public int getMaxDistance() {
		return maxDistance;
	}

	@NotNull
	@Override
	public String getSymbol() {
		return colour.getColourCode() + symbol;
	}

	@NotNull
	public Colour getColour() {
		return colour;
	}

	public void setColour(@NotNull Colour colour) {
		this.colour = colour;
	}

	public String getSymbol2() {
		return symbol;
	}

	@NotNull
	public Direction[] getValidDirections() {
		return validDirections;
	}

	@Override
	public int hashCode() {
		int result = maxDistance;
		result = 31 * result + colour.hashCode();
		result = 31 * result + symbol.hashCode();
		result = 31 * result + Arrays.hashCode(validDirections);
		return result;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Piece)) return false;

		Piece piece = (Piece) o;

		if (maxDistance != piece.maxDistance) return false;
		if (colour != piece.colour) return false;
		if (!symbol.equals(piece.symbol)) return false;
		// Probably incorrect - comparing Object[] arrays with Arrays.equals
		return Arrays.equals(validDirections, piece.validDirections);
	}

	@Override
	public String toString() {
		return "Piece{" +
				"maxDistance=" + maxDistance +
				", colour=" + colour +
				", symbol='" + symbol + '\'' +
				", validDirections=" + Arrays.toString(validDirections) +
				'}';
	}
}

File: Base/src/main/kotlin/macro303/base/pieces/Piece.kt
package macro303.base.pieces

import macro303.base.Board
import macro303.base.Colour
import macro303.base.Direction
import macro303.base.IBoard
import org.apache.logging.log4j.LogManager
import java.util.*

/**
 * Created by Macro303 on 2017-10-20.
 */
abstract class Piece(
	var teamColour: Colour,
	val maxDistance: Int,
	val validDirections: Array<Direction>
) {
	abstract val symbol: String

	open fun validMovement(start: Pair<Int, Int>, end: Pair<Int, Int>, board: Board): Boolean {
		val valid = validDirections.contains(
			IBoard.calculateDirection(
				start = start,
				end = end
			)
		) && maxDistance >= IBoard.calculateDistance(start = start, end = end)
		LOGGER.trace("boolean validMovement(Pair<Int, Int>, Pair<Int, Int>, Board) = $valid")
		return valid
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Piece) return false

		if (teamColour != other.teamColour) return false
		if (maxDistance != other.maxDistance) return false
		if (!Arrays.equals(validDirections, other.validDirections)) return false

		return true
	}

	override fun hashCode(): Int {
		var result = teamColour.hashCode()
		result = 31 * result + maxDistance
		result = 31 * result + Arrays.hashCode(validDirections)
		return result
	}

	override fun toString(): String {
		return "Piece(teamColour=$teamColour, maxDistance=$maxDistance, validDirections=${Arrays.toString(
			validDirections
		)})"
	}

	companion object {
		private val LOGGER = LogManager.getLogger(Piece::class.java)
	}
}